### PR TITLE
Support new thread and message attributes (position, message_count, total_message_sent)

### LINKED ISSUE #640: Support new thread and message attributes (position, message_count, total_message_sent)
### Summary

https://github.com/discord/discord-api-docs/pull/5206

### What is the feature request for?

The core library

### The Feature

`Message.position`, should probably default to None outside of forum channels, or we could think about using `-1` so its always an integer.
`Thread.message_count`
`Thread.total_message_sent`

However, there is a small implementation detail here: 
Should we increment the `message_count` and `total_message_sent` variables when we receive a `message_create` event for the thread? I believe we should.

In addition, when we receive a message_delete event in a thread, we should decrement `message_count` if the thread is cached whether or not the message is cached.

### PR DESCRIPTION
**Pull request by Snipy7374**
## Summary

This PR closes #640 

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/threads.py [hunk 1]
```diff
@@ -193,6 +206,7 @@ def _from_data(self, data: ThreadPayload):
         self.last_message_id = _get_as_snowflake(data, "last_message_id")
         self.slowmode_delay = data.get("rate_limit_per_user", 0)
         self.message_count = data.get("message_count")
+        self.total_message_sent = data.get("total_message_sent")
```
Comment (onerandomusername) [inline disnake/threads.py:209]: in theory, both of these could result in message_count and total_message_sent being `None`.

File: disnake/types/channel.py [hunk 2]
```diff
@@ -138,7 +140,8 @@ class ThreadChannel(_BaseChannel, _ThreadChannelOptional):
     nsfw: bool
     last_message_id: Optional[Snowflake]
     rate_limit_per_user: int
-    message_count: int
+    message_count: Optional[int]
+    total_message_sent: Optional[int]
```
Comment (onerandomusername) [inline disnake/types/channel.py:144]: @shiftinv are we sure these are optional?

File: disnake/types/channel.py [hunk 3]
```diff
@@ -147,6 +150,8 @@ class _ForumChannelOptional(TypedDict, total=False):
     topic: Optional[str]
     last_message_id: Optional[Snowflake]
     default_auto_archive_duration: ThreadArchiveDurationLiteral
+    message_count: Optional[int]
+    total_message_sent: Optional[int]
```
Comment (onerandomusername) [inline disnake/types/channel.py:154]: these are not part of forum channels

File: disnake/state.py [hunk 4]
```diff
@@ -777,12 +778,19 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
...
 
     def parse_message_delete(self, data: gateway.MessageDeleteEvent) -> None:
         raw = RawMessageDeleteEvent(data)
         found = self._get_message(raw.message_id)
         raw.cached_message = found
         self.dispatch("raw_message_delete", raw)
+        guild = self._get_guild(raw.guild_id)
+        thread = guild and guild.get_thread(raw.channel_id)
+        if thread:
+            thread.message_count -= 1
```
Comment (onerandomusername) [inline disnake/state.py:793]: it is possible for this to be None (and this goes for all of the other uses in this file) which would cause a TypeError when trying to subtract one, but please wait for resolving if this should actually be optional, per a different comment concerning the typings.

Additionally, this may have issues with threads older than july and result in a negative message_count, although it may not be worth catching that rare case...

File: disnake/state.py [hunk 5]
```diff
@@ -797,6 +805,13 @@ def parse_message_delete_bulk(self, data: gateway.MessageDeleteBulkEvent) -> Non
             found_messages = []
         raw.cached_messages = found_messages
         self.dispatch("raw_bulk_message_delete", raw)
+        guild = self._get_guild(raw.guild_id)
+        thread = guild and guild.get_thread(raw.channel_id)
+        for _ in found_messages:
+            if thread:
+                thread.message_count -= 1
+            else:
+                continue
```
Comment (onerandomusername) [inline disnake/state.py:814]: They don't all have to be cached, and bulk deletes all occur in the same channel. As such, we can subtract `len(data['ids'])` from the thread without having a for-loop.

```suggestion
        guild = self._get_guild(raw.guild_id)
        thread = guild and guild.get_thread(raw.channel_id)
        if thread:
            thread.message_count -= len(raw.message_ids)
```

File: changelog/640.feature.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Adds :attr:`Thread.message_count`, :attr:`Thread.total_message_sent` and :attr:`Message.position` attributes.
```
Comment (onerandomusername) [inline changelog/640.feature.rst:1]:
```suggestion
Add :attr:`Thread.message_count`, :attr:`Thread.total_message_sent` and :attr:`Message.position` attributes.
```

**Review by onerandomusername — CHANGES_REQUESTED** [review]
Great work so far!

**Comment by onerandomusername** [comment]
Discovered an API bug while testing this issue: 
- https://github.com/discord/discord-api-docs/issues/5436

File: disnake/state.py (hunk 4 above)
Comment (Snipy7374) [reply disnake/state.py:793]: For the second part of your comment:

maybe we can check if the `created_at` property of the thread is older than July 2, 2022 23:59:59, if it is we ignore it, what do you think? maybe there's other ways

File: disnake/types/channel.py (hunk 3 above)
Comment (Snipy7374) [inline disnake/types/channel.py:154]:
```suggestion
```

File: disnake/types/channel.py [hunk 7]
```diff
@@ -147,6 +150,8 @@ class _ForumChannelOptional(TypedDict, total=False):
     topic: Optional[str]
     last_message_id: Optional[Snowflake]
     default_auto_archive_duration: ThreadArchiveDurationLiteral
+    message_count: Optional[int]
```
Comment (Snipy7374) [inline disnake/types/channel.py:153]:
```suggestion
```

File: disnake/state.py [hunk 8]
```diff
@@ -777,12 +778,19 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
         # we ensure that the channel is a type that implements last_message_id
         if channel and channel.__class__ in (TextChannel, Thread, VoiceChannel):
             channel.last_message_id = message.id  # type: ignore
+        if channel and isinstance(channel, Thread):
+            channel.total_message_sent += 1
+            channel.message_count += 1
```
Comment (shiftinv) [inline disnake/state.py:783]: This should check `message.type is MessageType.thread_starter_message` and not increment the counter in that case, to mirror the API behavior:
(lib incremented counters, API didn't)
image

[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/state.py:783]: First one looks good, I don't think the second one is necessary as the overhead of the check should be close to none.

File: disnake/types/channel.py (hunk 2 above)
Comment (shiftinv) [reply disnake/types/channel.py:144]: The other thread model type assumes they are always set, which may be a valid assumption but I'd rather not take any chances.
The definitions in both types should be in their respective `_Optional` parents, and perhaps default to `0` when parsing?

Comment (shiftinv) [reply disnake/types/channel.py:144]: Just noticed that at least in this case they're already part of `_ThreadChannelOptional`, so they should be removed from `ThreadChannel` anyway

File: disnake/state.py (hunk 4 above)
Comment (shiftinv) [reply disnake/state.py:793]: > this may have issues with threads older than july and result in a negative message_count

`thread.message_count = max(0, thread.message_count - 1)` would avoid that

File: disnake/threads.py (hunk 1 above)
Comment (Snipy7374) [reply disnake/threads.py:209]: > in theory, both of these could result in message_count and total_message_sent being `None`.

@onerandomusername so should we suppose if they're `None` they're 0?

File: disnake/state.py (hunk 8 above)
Comment (Snipy7374) [reply disnake/state.py]: > This should check `message.type is MessageType.thread_starter_message` and not increment the counter in that case, to mirror the API behavior: (lib incremented counters, API didn't) image

`Thread.total_message_sent` should count the starter message so i've applied the changes only to `Thread.message_count`, lemme know if that's right

Comment (shiftinv) [reply disnake/state.py]: I wasn't sure about that part, but it doesn't seem like that's the case - `total_message_sent` also shouldn't count the starter message:


<details>
<summary>code</summary>

```py
# create thread from message
msg = await ctx.send("test")
t_created = await msg.create_thread(name="h")
t_id = t_created.id

# check values in cache
t_cache = ctx.guild.get_thread(t_id)
print("cache:", t_cache.message_count, t_cache.total_message_sent)

# check values in api
t_fetch = await ctx.guild.fetch_channel(t_id)
print("fetch:", t_fetch.message_count, t_fetch.total_message_sent)
```

</details>

File: disnake/state.py [hunk 9]
```diff
@@ -786,7 +786,7 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
         if channel:
             if channel.__class__ in (TextChannel, Thread, VoiceChannel):
                 channel.last_message_id = message.id  # type: ignore
-            if channel.__class__ is Thread:
+            if type(channel) is Thread and message.type is not MessageType.thread_starter_message:
```
Comment (onerandomusername) [inline disnake/state.py:789]: this should be `channel.__class__` to be consistent with the line above

File: disnake/types/channel.py [hunk 10]
```diff
@@ -140,8 +140,6 @@ class ThreadChannel(_BaseChannel, _ThreadChannelOptional):
     nsfw: bool
     last_message_id: Optional[Snowflake]
     rate_limit_per_user: int
-    message_count: Optional[int]
-    total_message_sent: Optional[int]
     member_count: int
```
Comment (onerandomusername) [inline disnake/types/channel.py:143]: @shiftinv why were these lines deleted? do we have the thread typings defined twice?

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/types/channel.py (hunk 10 above)
Comment (Snipy7374) [reply disnake/types/channel.py]: > @shiftinv why were these lines deleted? do we have the thread typings defined twice?

If they're already in `_ThreadChannelOptional` we don't need them in `ThreadChannel` too no?

File: disnake/state.py (hunk 9 above)
Comment (Snipy7374) [reply disnake/state.py]: > this should be `channel.__class__` to be consistent with the line above

I was using `type()` coz pyright cry with `.__class__`

File: disnake/state.py [hunk 11]
```diff
@@ -782,14 +783,22 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
...
             self._messages.append(message)
         # we ensure that the channel is a type that implements last_message_id
-        if channel and channel.__class__ in (TextChannel, Thread, VoiceChannel):
-            channel.last_message_id = message.id  # type: ignore
+        if channel:
+            if channel.__class__ in (TextChannel, Thread, VoiceChannel):
+                channel.last_message_id = message.id  # type: ignore
+            if channel.__class__ is Thread and message.type is not MessageType.thread_starter_message:
+                channel.total_message_sent += 1
+                channel.message_count += 1
```
Comment (Snipy7374) [inline disnake/state.py:791]: @onerandomusername using `type()` in both ifs conditions we could remove the `# type: ignore`

Comment (shiftinv) [reply disnake/state.py:791]: The difference between `x.__class__` and `type(x)` is close to none at that scale (difference of a few nanoseconds), so let's go with `type(x)` here as pyright can properly narrow the type that way.

File: disnake/threads.py [hunk 12]
```diff
@@ -111,9 +111,21 @@ class Thread(Messageable, Hashable):
         Bots and users with :attr:`~Permissions.manage_channels` or
         :attr:`~Permissions.manage_messages` bypass slowmode.
     message_count: Optional[:class:`int`]
-        An approximate number of messages in this thread. This caps at 50.
+        An approximate number of messages in this thread.
+
+        .. note::
+
+            If the thread was created before July 1, 2022, this stops counting at 50.
```
Comment (onerandomusername) [inline disnake/threads.py:118]: take a look at discord/discord-api-docs#5437 for the wording here

File: disnake/threads.py [hunk 13]
```diff
@@ -111,9 +111,21 @@ class Thread(Messageable, Hashable):
...
     member_count: Optional[:class:`int`]
         An approximate number of members in this thread. This caps at 50.
+    total_message_sent: Optional[:class:`int`]
+        The total number of messages sent in the thread, including deleted messages.
+
+        .. versionadded:: 2.6
+
+        .. note::
+
+            If the thread was created before July 1, 2022, this stops counting at 50.
```
Comment (onerandomusername) [inline disnake/threads.py:128]: take a look at discord/discord-api-docs#5437 for the wording here

File: disnake/state.py [hunk 14]
```diff
@@ -782,14 +783,22 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
         if self._messages is not None:
             self._messages.append(message)
         # we ensure that the channel is a type that implements last_message_id
-        if channel and channel.__class__ in (TextChannel, Thread, VoiceChannel):
-            channel.last_message_id = message.id  # type: ignore
+        if channel:
+            if channel.__class__ in (TextChannel, Thread, VoiceChannel):
+                channel.last_message_id = message.id  # type: ignore
+            if channel.__class__ is Thread and message.type is not MessageType.thread_starter_message:
+                channel.total_message_sent += 1
```
Comment (onerandomusername) [inline disnake/state.py:790]: missing a `# type: ignore`

File: disnake/threads.py (hunk 12 above)
Comment (Snipy7374) [reply disnake/threads.py:118]: > take a look at discord/discord-api-docs#5437 for the wording here

ok I'll fix it

File: disnake/state.py (hunk 14 above)
Comment (Snipy7374) [reply disnake/state.py:790]: > missing a `# type: ignore`

so in conclusion what should I use? `.__class__`, `type()`, `isinstance()`, `typing.cast()` or just `# type: ignore`?

File: disnake/threads.py [hunk 15]
```diff
@@ -192,7 +205,8 @@ def _from_data(self, data: ThreadPayload):
         self._type: ThreadType = try_enum(ChannelType, data["type"])  # type: ignore
         self.last_message_id = _get_as_snowflake(data, "last_message_id")
         self.slowmode_delay = data.get("rate_limit_per_user", 0)
-        self.message_count = data.get("message_count")
+        self.message_count = data.get("message_count", 0)
+        self.total_message_sent = data.get("total_message_sent", 0)
```
Comment (shiftinv) [inline disnake/threads.py:209]:
```suggestion
        self.message_count = data.get("message_count") or 0
        self.total_message_sent = data.get("total_message_sent") or 0
```
Almost the same thing, but falls back to `0` if `data["message_count"]` is `None` (which shouldn't happen, but I'd rather play it safe since these are being incremented in the gateway handlers)

**Review by shiftinv — APPROVED** [review]
Looks good to me, thank you!
Tested it in forum channels as well, seems to work fine in all cases.

**Review by onerandomusername — APPROVED** [review]
